src/storage/index.js

```diff
@@ -452,4 +452,82 @@ export function deleteFile(path: string) {
 }
 
+/**
+ * Loop over the list of files in a Gaia hub, and run a callback on each entry.
+ * Not meant to be called by external clients.
+ * @param {GaiaHubConfig} hubConfig - the Gaia hub config
+ * @param {String | null} page - the page ID
+ * @param {number} callCount - the loop count
+ * @param {number} fileCount - the number of files listed so far
+ * @param {function} callback - the callback to invoke on each file.  If it returns a falsey
+ *  value, then the loop stops.  If it returns a truthy value, the loop continues.
+ * @returns {Promise} that resolves to the number of files listed.
+ */
+function listFilesLoop(hubConfig: GaiaHubConfig,
+                       page: string | null,
+                       callCount: number,
+                       fileCount: number,
+                       callback: (name: string) => boolean) : Promise<number> {
+  if (callCount > 65536) {
+    // this is ridiculously huge, and probably indicates
+    // a faulty Gaia hub anyway (e.g. on that serves endless data)
+    throw new Error('Too many entries to list')
+  }
+
+  let httpStatus
+  const pageRequest = JSON.stringify({ page })
+  return fetch(`${hubConfig.server}/list-files/${hubConfig.address}`,
+    {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        'Content-Length': `${pageRequest.length}`,
+        Authorization: `bearer ${hubConfig.token}`
+      },
+      body: pageRequest
+    })
+    .then((response) => {
+      httpStatus = response.status
+      if (httpStatus >= 400) {
+        throw new Error(`listFiles failed with HTTP status ${httpStatus}`)
+      }
+      return response.text()
+    })
+    .then((responseText) => JSON.parse(responseText))
+    .then((responseJSON) => {
+      const entries = responseJSON.entries
+      const nextPage = responseJSON.page
+      if (entries === null || entries === undefined) {
+        // indicates a misbehaving Gaia hub or a misbehaving driver
+        // (i.e. the data is malformed)
+        throw new Error('Bad listFiles response: no entries')
+      }
+      for (let i = 0; i < entries.length; i++) {
+        const rc = callback(entries[i])
+        if (!rc) {
+          // callback indicates that we're done
+          return Promise.resolve(fileCount + i)
+        }
+      }
+      if (nextPage && entries.length > 0) {
+        // keep going -- have more entries
+        return listFilesLoop(hubConfig, nextPage, callCount + 1,
+            fileCount + entries.length, callback)
+      } else {
+        // no more entries -- end of data
+        return Promise.resolve(fileCount)
+      }
+    })
+}
+
+/**
+ * List the set of files in this application's Gaia storage bucket.
+ * @param {function} callback - a callback to invoke on each named file
+ * @return {Promise} that resolves to the number of files listed
+ */
+export function listFiles(callback: (name: string) => boolean) : Promise<number> {
```

Comment:
I'm not entirely clear on the purpose of the callback to invoke on each file.

As a developer using this API, I would expect it to return a list of files instead of taking a callback that performs an operation on the files and returning the number of files.

It seems like this is actually couple api calls in one:

`listFiles` -> this could optionally take: a number of files to list, a page number and/or potentially a query string and resolve to an array of the file paths. Each of 

`totalFiles` or `countFiles` -> this could resolve to the number of files and is just a convenience method implemented by calling listFiles and returning the length of the array

`forEach`- applying a function to each element of the array from `listFiles` - this already exists in JavaScript https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

What about changing the `listFiles` function signature to be something like the following:

`function listFiles(query: string = '*', maxResults: number = 1000, page: number = 0): Promise<Array<string>>`

This would allow developers who want to do something like simple, like list all the files in a bucket that doesn't have that many files to do:

`blockstack.listFiles().then((files) => { // do something with files here } `

Reply:
> As a developer using this API, I would expect it to return a list of files instead of taking a callback that performs an operation on the files and returning the number of files.

The problem is that this list can be huge, and/or can take a while to assemble if either the Gaia hub, the backing store, or the user's network connection is slow.  This API gives developers the ability to work with partial results while the next pages of the listing are in flight.

> this could optionally take: a number of files to list, a page number and/or potentially a query string and resolve to an array of the file paths. Each of

A few things:
* Pages aren't numbers.  Usually, they're HMAC'ed database cursors of some kind, and are opaque to us.  The page identifier is specific to the Gaia driver, and page identifiers cannot be assumed to be reusable across multiple requests.  Also, most backing stores we support do not have the notion of an "offset" at which to start listing---you have to start at the beginning, and iterate through the list of files.
* If you want to list a certain number of files, this can already be serviced with the existing API---simply have your callback keep count of the number of files listed so far.
* Different backing stores will have different query strings (they may not even support SQL), so it doesn't make sense to allow the application to pass this information in.  We don't want applications to get coupled to particular backing stores.

> totalFiles or countFiles -> this could resolve to the number of files and is just a convenience method implemented by calling listFiles and returning the length of the array

Not all backing stores support a way to efficiently query the number of files.  Also, we'd need to add a separate API route in Gaia to support this.  I think that if we want to have something like `countFiles`, we should discuss it first on Gaia.

> This would allow developers who want to do something like simple, like list all the files in a bucket that doesn't have that many files to do:

This is also achievable with the given API.  Simply have the callback accumulate all file names.

Reply:
Okay. This makes sense to me. Let's go with this API and if developers find it confusing we revisit it or create convenience methods for various specific use cases.